File: src/lib/json/jsonqueryparser.cpp
/* 
 * File:   JSONQueryParser.cpp
 * Author: Thomas Krause <thomaskrause@posteo.de>
 * 
 * Created on 3. Januar 2016, 16:05
 */

#include <annis/json/jsonqueryparser.h>
#include <annis/annosearch/exactannovaluesearch.h>
#include <annis/annosearch/exactannokeysearch.h>
#include <annis/annosearch/regexannosearch.h>
#include <annis/operators/precedence.h>
#include <annis/operators/dominance.h>
#include <annis/operators/pointing.h>
#include <annis/operators/inclusion.h>
#include <annis/operators/overlap.h>
#include <annis/operators/identicalcoverage.h>
#include <map>

using namespace annis;

JSONQueryParser::JSONQueryParser()
{
}

std::shared_ptr<Query> JSONQueryParser::parse(const DB& db, std::istream& jsonStream, bool optimize)
{
  std::shared_ptr<Query> q = std::make_shared<Query>(db, optimize);

  // parse root as value
  Json::Value root;
  jsonStream >> root;

  // get the first alternative (we don't support more than one currently)
  const auto& alternatives = root["alternatives"];
  if (alternatives.size() != 0)
  {
    const auto& firstAlt = alternatives[0];

    // add all nodes
    const auto& nodes = firstAlt["nodes"];

    std::map<std::uint64_t, size_t> nodeIdToPos;
    for (auto it = nodes.begin(); it != nodes.end(); it++)
    {
      auto& n = *it;
      nodeIdToPos[std::stoull(it.name())] = parseNode(db, n, q);
    }

    // add all joins
    const auto& joins = firstAlt["joins"];
    for (auto it = joins.begin(); it != joins.end(); it++)
    {
      parseJoin(db, *it, q, nodeIdToPos);
    }


  }
  return q;
}

size_t JSONQueryParser::parseNode(const DB& db, const Json::Value node, std::shared_ptr<Query> q)
{

  // annotation search?
  if (node["nodeAnnotations"].isArray() && node["nodeAnnotations"].size() > 0)
  {
    // get the first one
    auto nodeAnno = node["nodeAnnotations"][0];

    return addNodeAnnotation(db, q, optStr(nodeAnno["namespace"]),
      optStr(nodeAnno["name"]), optStr(nodeAnno["value"]),
      optStr(nodeAnno["textMatching"]));

  }// end if annotation search
  else
  {
    // check for special non-annotation search constructs
    // token search?
    if (node["spannedText"].isString()
      || (node["token"].isBool() && node["token"].asBool()))
    {
      return addNodeAnnotation(db, q, optStr(annis_ns), optStr(annis_tok),
        optStr(node["spannedText"]),
        optStr(node["spanTextMatching"]), true);
    }// end if token has spanned text
    else
    {
      // just search for any node
      return addNodeAnnotation(db, q, optStr(annis_ns), optStr(annis_node_name),
        optStr(), optStr());
    }
  } // end if special case

}

size_t JSONQueryParser::addNodeAnnotation(const DB& db,
  std::shared_ptr<Query> q,
  boost::optional<std::string> ns,
  boost::optional<std::string> name,
  boost::optional<std::string> value,
  boost::optional<std::string> textMatching,
  bool wrapEmptyAnno)
{

  if (value)
  {
    bool exact = *textMatching == "EXACT_EQUAL";
    bool regex = *textMatching == "REGEXP_EQUAL";
    if(regex)
    {
      if(canReplaceRegex(*value))
      {
        exact = true;
      }
    }
    
    // search for the value
    if (exact)
    {
      // has namespace?
      if (ns)
      {
        return q->addNode(std::make_shared<ExactAnnoValueSearch>(db,
          *ns,
          *name,
          *value),
          wrapEmptyAnno);
      }
      else
      {
        return q->addNode(std::make_shared<ExactAnnoValueSearch>(db,
          *name,
          *value),
          wrapEmptyAnno);
      }
    }
    else if (regex)
    {
      // has namespace?
      if (ns)
      {
        return q->addNode(std::make_shared<RegexAnnoSearch>(db,
          *ns,
          *name,
          *value),
          wrapEmptyAnno);
      }
      else
      {
        return q->addNode(std::make_shared<RegexAnnoSearch>(db,
          *name,
          *value),
          wrapEmptyAnno);
      }
    }

  }// end if has value
  else
  {
    // only search for key
    // has namespace?
    if (ns)
    {
      return q->addNode(std::make_shared<ExactAnnoKeySearch>(db,
        *ns,
        *name),
        wrapEmptyAnno);
    }
    else
    {
      return q->addNode(std::make_shared<ExactAnnoKeySearch>(db,
        *name),
        wrapEmptyAnno);
    }
  }
  // should never happen
  assert(false);
}

void JSONQueryParser::parseJoin(const DB& db, const Json::Value join, std::shared_ptr<Query> q,
  const std::map<std::uint64_t, size_t>& nodeIdToPos)
{
  // get left and right index
  if (join["left"].isUInt64() && join["right"].isUInt64())
  {
    auto leftID = join["left"].asUInt64();
    auto rightID = join["right"].asUInt64();

    auto itLeft = nodeIdToPos.find(leftID);
    auto itRight = nodeIdToPos.find(rightID);

    if (itLeft != nodeIdToPos.end() && itRight != nodeIdToPos.end())
    {

      auto op = join["op"].asString();
      if (op == "Precedence")
      {
        auto minDist = join["minDistance"].asUInt();
        auto maxDist = join["maxDistance"].asUInt();
        q->addOperator(std::make_shared<Precedence>(db,
          minDist, maxDist),
          itLeft->second, itRight->second);
      }
      else if (op == "Inclusion")
      {
        q->addOperator(std::make_shared<Inclusion>(db), itLeft->second, itRight->second);
      }
      else if (op == "Overlap")
      {
        q->addOperator(std::make_shared<Overlap>(db), itLeft->second, itRight->second);
      }
      else if (op == "IdenticalCoverage")
      {
        q->addOperator(std::make_shared<IdenticalCoverage>(db), itLeft->second, itRight->second);
      }
      else if (op == "Dominance")
      {

        std::string name = join["name"].isString() ? join["name"].asString() : "";

        if (join["edgeAnnotations"].isArray() && join["edgeAnnotations"].size() > 0)
        {
          auto anno = getEdgeAnno(db, join["edgeAnnotations"][0]);
          q->addOperator(std::make_shared<Dominance>(db, "", name, anno),
            itLeft->second, itRight->second);

        }
        else
        {

          auto minDist = join["minDistance"].asUInt();
          auto maxDist = join["maxDistance"].asUInt();
          if (minDist == 0 && maxDist == 0)
          {
            // unlimited range
            minDist = 1;
            maxDist = uintmax;
          }

          q->addOperator(std::make_shared<Dominance>(db,
            "", name,
            minDist, maxDist),
            itLeft->second, itRight->second);
        }
      }
      else if (op == "Pointing")
      {

        std::string name = join["name"].asString();

        if (join["edgeAnnotations"].isArray() && join["edgeAnnotations"].size() > 0)
        {
          auto anno = getEdgeAnno(db, join["edgeAnnotations"][0]);
          q->addOperator(std::make_shared<Pointing>(db, "", name, anno),
            itLeft->second, itRight->second);

        }
        else
        {

          auto minDist = join["minDistance"].asUInt();
          auto maxDist = join["maxDistance"].asUInt();

          if (minDist == 0 && maxDist == 0)
          {
            // unlimited range
            minDist = 1;
            maxDist = uintmax;
          }

          q->addOperator(std::make_shared<Pointing>(db,
            "", name, minDist, maxDist),
            itLeft->second, itRight->second);
        }
      }

    }

  }
}

Annotation JSONQueryParser::getEdgeAnno(const DB& db, const Json::Value& edgeAnno)
{

  std::uint32_t ns = 0;
  std::uint32_t name = 0;
  std::uint32_t value = 0;

  if (edgeAnno["textMatching"].asString() == "EXACT_EQUAL")
  {
    if (edgeAnno["namespace"].isString())
    {
      std::string nsStr = edgeAnno["namespace"].asString();
      auto search = db.strings.findID(nsStr);
      if (search.first)
      {
        ns = search.second;
      }
    }
    if (edgeAnno["name"].isString())
    {
      std::string nameStr = edgeAnno["name"].asString();
      auto search = db.strings.findID(nameStr);
      if (search.first)
      {
        name = search.second;
      }
    }
    if (edgeAnno["value"].isString())
    {
      std::string valueStr = edgeAnno["value"].asString();
      auto search = db.strings.findID(valueStr);
      if (search.first)
      {
        value = search.second;
      }
    }
  }
  // TODO: what about regex?

  return Init::initAnnotation(name, value, ns);
}

bool JSONQueryParser::canReplaceRegex(const std::string& str) 
{
  // Characters that have a meaning according to
  // https://github.com/google/re2/wiki/Syntax
  // Characters used in not supported functions are not included.
  if(str.find_first_of(".[]\\|*+?{}()^$") == std::string::npos)
  {
    // No meta character found in string, might be replaced    
    RE2 regex(str);
    if(regex.ok())
    {
      return true;
    }
    else
    {
      // If there is an error during parsing this is still a regex (an invalid one).
      // Treating it like a exact string would not give the same result.
      return false;
    }
  }
  else
  {
    // contains special regex characters
    return false;
  }
}


JSONQueryParser::~JSONQueryParser()
{
}

File: graphannis-rs/src/parser/jsonqueryparser.rs
use serde_json;
use query::conjunction::Conjunction;
use query::disjunction::Disjunction;
use exec::nodesearch::NodeSearchSpec;

use graphdb::{ANNIS_NS, TOK};

use std::collections::BTreeMap;

pub fn parse(query_as_string: &str) -> Option<Disjunction> {
    let root : serde_json::Value = serde_json::from_str(query_as_string).ok()?;

    let mut conjunctions: Vec<Conjunction> = Vec::new();
    // iterate over all alternatives
    let alternatives = root["alternatives"].as_array()?;
    
    for alt in alternatives.iter() {
        let mut q = Conjunction::new();

        // add all nodes
        let mut node_id_to_pos: BTreeMap<usize, usize> = BTreeMap::new();
        if let serde_json::Value::Object(ref nodes) = alt["nodes"] {
            for (node_name, node) in nodes.iter() {
                if let Some(node_obj) = node.as_object() {
                    if let Ok(ref node_id) = node_name.parse::<u64>() {
                        let node_id = node_id.clone() as usize;

                        let pos = parse_node(node_obj, &mut q);
                        node_id_to_pos.insert(node_id, pos);
                    }
                }
            }
        }

        // TODO: add all joins
        if let serde_json::Value::Array(ref joins) = alt["joins"] {
            for j in joins.iter() {
                if let &serde_json::Value::Object(ref j_obj) = j {
                    parse_join(j_obj, &mut q, &node_id_to_pos);
                }
            }
        }

        // TODO: add all meta-data

        conjunctions.push(q);
        unimplemented!();
    }



    if !conjunctions.is_empty() {
        return Some(Disjunction::new(conjunctions));
    }
    
    return None;
}

fn parse_node(node: &serde_json::Map<String, serde_json::Value>, q: &mut Conjunction) -> usize {
    // annotation search?
    if let serde_json::Value::Array(ref a) = node["nodeAnnotations"] {
        if !a.is_empty() {
            // get the first one
            let a = &a[0];
            return add_node_annotation(
                q,
                a["namespace"].as_str(),
                a["name"].as_str(),
                a["value"].as_str(),
                is_regex(a),
            );
        }
    }

    // check for special non-annotation search constructs
    // token search?
    if node["spannedText"].is_string()
        || (node["token"].is_boolean() && node["token"].is_boolean()) {
        let spanned = node["spannedText"].as_str();

        let mut leafs_only = false;
        if let Some(is_token) = node["token"].as_bool() {
            if is_token {
                // special treatment for explicit searches for token (tok="...)
                leafs_only = true;
            }
        }

        if let Some(tok_val) = spanned {
            if node["textMatching"].as_str() == Some("REGEXP_EQUAL") {
                return q.add_node(NodeSearchSpec::RegexTokenValue{val: String::from(tok_val), leafs_only,});
            } else {
                return q.add_node(NodeSearchSpec::ExactTokenValue{val: String::from(tok_val), leafs_only,});
            }
        } else {
            return q.add_node(NodeSearchSpec::AnyToken);
        }


    } else {
        // just search for any node
        return q.add_node((NodeSearchSpec::AnyNode));
    }
}

fn parse_join(join: &serde_json::Map<String, serde_json::Value>, q: &mut Conjunction, node_id_to_pos: &BTreeMap<usize, usize>) -> usize { 
    // get left and right index
    if let (Some(left_id), Some(right_id)) = (join["left"].as_u64(), join["right"].as_u64()) {
        let left_id = left_id as usize;
        let right_id = right_id as usize;
        if let (Some(pos_left),Some(pos_right)) = (node_id_to_pos.get(&left_id),node_id_to_pos.get(&right_id)) {
            if let Some(op) = join["op"].as_str() {

            }
        }
    }
    
    unimplemented!()
}

fn is_regex(json_node : &serde_json::Value) -> bool {
    if let Some(tm) = json_node["textMatching"].as_str() {
        if tm == "REGEXP_EQUAL" {
            return true;
        }
    }
    return false;
}

fn add_node_annotation(
    q: &mut Conjunction,
    ns: Option<&str>,
    name: Option<&str>,
    value: Option<&str>,
    regex : bool,
) -> usize {
    if let Some(name_val) = name {
        // TODO: replace regex with normal text matching if this is not an actual regular expression

        // search for the value
        if regex {
            if let Some(val) = value {
                let mut n: NodeSearchSpec =
                    NodeSearchSpec::new_regex(ns, name_val, val);
                return q.add_node(n);
            }
        } else  {
            // has namespace?
            let mut n: NodeSearchSpec =
                NodeSearchSpec::new_exact(ns, name_val, value);
            return q.add_node(n);
        }
    }
    unimplemented!()
}
